### DISCUSSION TITLE
the material property import from .CSV file is not match [.C file problem]

### DISCUSSION THREAD

**Discussion by Traiwit** [idaholab/moose#18880] [Q&A General] [answered]
Hi all, 

I want to introduce a new material property (Damage) base on the element ID

I use `ElementPropertyReadFile` to read data from csv file, 

```
  [./damage_read]
    type = ElementPropertyReadFile
    prop_file_name = 'damage.csv'
    nprop = 1
    read_type = element
  [../]
```

but they don't match, I'm pretty sure my .C file isn't correct. could you please help me check what did I do wrong?

This is my .C file, I didn't know where to start, so I modified `GenericConstantArray.C`

```
InputParameters
GenericConstantArrayTC::validParams()
{

  InputParameters params = Material::validParams();
  params.addRequiredParam<std::string>("prop_name",
                                       "The name of the property this material will have");
  // params.addParam<RealEigenVector>("prop_value",
  //                                          "The values associated with the named property");
  // params.declareControllable("prop_value");
  params.addClassDescription(
      "A material evaluating one material property in type of RealEigenVector");
  params.set<MooseEnum>("constant_on") = "SUBDOMAIN";
  params.addParam<UserObjectName>("read_prop_user_object",
                                      "The ElementReadPropertyFile "
                                      "GeneralUserObject to read element "
                                      "specific property values from file");
  return params;
}
GenericConstantArrayTC::GenericConstantArrayTC(const InputParameters & parameters)
  : Material(parameters),
    _prop_name(getParam<std::string>("prop_name")),
    // _prop_value(getParam<RealEigenVector>("prop_value")),
    _property(declareProperty<Real>(_prop_name)),
    _mat_prop(declareProperty<Real>("prop_value")),
    _read_prop_user_object(isParamValid("read_prop_user_object")
                               ? &getUserObject<ElementPropertyReadFile>("read_prop_user_object")
                               : nullptr)
{
}

void
GenericConstantArrayTC::initQpStatefulProperties()
{
  computeQpProperties();
}

void
GenericConstantArrayTC::computeQpProperties()
{
   _mat_prop[_qp] = _read_prop_user_object->getData(_current_elem, 0);
  _property[_qp] = _mat_prop[_qp];
}
```

** I think the problem is from the last 2 line **

Basically the data read from csv file is store at _read_prop_user_object then > _mat_prop then > property

Validation: I tested by let damage = elementID

![image](https://user-images.githubusercontent.com/79562439/134155592-c85b2e40-8aa2-4baf-be3d-72468a4cd9d4.png)

However, the result isn't that I wanted, the elementID should be equal to the damage

![image](https://user-images.githubusercontent.com/79562439/134155784-40554364-2035-4767-8e30-5e4f374bd968.png)

Thank you very much!

Kind regards,
Trai

**Comment by GiudGiud** [comment, marked as answer]
Hello

There's a few issues with that code. 
-If you forget the read_user_object parameter, then you ll be trying to access a nullptr and that will segfault. 

-Why do you have both mat_prop and property ? You only need one. 

-   params.set<MooseEnum>("constant_on") = "SUBDOMAIN";
This is going to mess with the evaluation, as it ll tell the code to evaluate it once per subdomain and not once per element / quadrature point. 

Guillaume

**Reply by Traiwit** [reply]
Hi @GiudGiud,

As you know my C++ isnt good, so i only follow the example and put things together.

> -If you forget the read_user_object parameter, then you ll be trying to access a nullptr and that will segfault.

should I just remove ': nullptr' from

```
 _read_prop_user_object(isParamValid("read_prop_user_object")
                               ? &getUserObject<ElementPropertyReadFile>("read_prop_user_object")
                               : nullptr)

```

> -Why do you have both mat_prop and property

so at the last section, should I just remove this line  `_property[_qp] = _mat_prop[_qp];` ?

> params.set("constant_on") = "SUBDOMAIN";

not sure what to do with this, do I just change subdomain to element?

Thank you for helping me again @GiudGiud :) 

Kind regards,
Traiwit

**Reply by GiudGiud** [reply]
1) that works
2) Remove _property and the declareProperty and remove it from the header
3) just delete that line params.set("constant_on") = "SUBDOMAIN";

**Comment by Traiwit** [comment]
all good now, thanks @GiudGiud !

![image](https://user-images.githubusercontent.com/79562439/134264723-9f875a60-e00e-4d00-b1b4-f3b33966ade7.png)
